packages/go/index.ts

```diff
@@ -94,332 +99,456 @@ export async function build({
-        break;
+    debug(`Parsing AST for "${entrypoint}"`);
+    let analyzed: string;
+    try {
+      const goModAbsPath = await findGoModPath(workPath);
+      if (goModAbsPath) {
+        debug(`Found ${goModAbsPath}"`);
       }
 
-      if (!isGoModInRootDir && config.zeroConfig && file === 'api/go.mod') {
-        // We didn't find `/go.mod` but we found `/api/go.mod` so move it to the root
+      analyzed = await getAnalyzedEntrypoint(
+        workPath,
+        entrypointAbsolute,
+        dirname(goModAbsPath)
+      );
+    } catch (err) {
+      console.log(`Failed to parse AST for "${entrypoint}"`);
+      throw err;
+    }
+
+    if (!analyzed) {
+      const err = new Error(
+        `Could not find an exported function in "${entrypoint}"
+  Learn more: https://vercel.com/docs/runtimes#official-runtimes/go
+        `
+      );
+      console.log(err.message);
+      throw err;
+    }
+
+    const parsedAnalyzed = JSON.parse(analyzed) as Analyzed;
+
+    // find `go.mod` in modFiles
+    const entrypointDirname = dirname(entrypointAbsolute);
+    let isGoModExist = false;
+    let goModPath = '';
+    let isGoModInRootDir = false;
+
+    const modFileRefs = await glob('**/*.mod', workPath);
+    const modFiles = Object.keys(modFileRefs);
+
+    for (const file of modFiles) {
+      const fileDirname = dirname(file);
+      if (file === 'go.mod') {
         isGoModExist = true;
         isGoModInRootDir = true;
-        goModPath = join(fileDirname, '..');
-        const pathParts = fsPath.split(sep);
-        pathParts.pop(); // Remove go.mod
-        pathParts.pop(); // Remove api
-        pathParts.push('go.mod');
-        const newFsPath = pathParts.join(sep);
-        debug(`Moving api/go.mod to root: ${fsPath} to ${newFsPath}`);
-        await move(fsPath, newFsPath);
-        const oldSumPath = join(dirname(fsPath), 'go.sum');
-        const newSumPath = join(dirname(newFsPath), 'go.sum');
-        if (await pathExists(oldSumPath)) {
-          debug(`Moving api/go.sum to root: ${oldSumPath} to ${newSumPath}`);
-          await move(oldSumPath, newSumPath);
+        goModPath = join(workPath, fileDirname);
+      } else if (file.endsWith('go.mod')) {
+        if (entrypointDirname === fileDirname) {
+          isGoModExist = true;
+          goModPath = join(workPath, fileDirname);
+
+          debug(`Found file dirname equals entrypoint dirname: ${fileDirname}`);
+          break;
+        }
+
+        if (!isGoModInRootDir && config.zeroConfig && file === 'api/go.mod') {
+          // We didn't find `/go.mod` but we found `/api/go.mod` so move it to the root
+          isGoModExist = true;
+          isGoModInRootDir = true;
+          goModPath = join(fileDirname, '..');
+          const pathParts = file.split(sep);
+          pathParts.pop(); // Remove go.mod
+          pathParts.pop(); // Remove api
+          pathParts.push('go.mod');
+
+          const newRoot = pathParts.join(sep);
```

Comment:
I understand your changes don't really change the method the path was rewritten, but I've run into cross platform issues with manually splitting on `path.sep`. I generally play it safe and just do:

```javascript
const newRoot = join(dirname(dirname(file)), 'go.mod');
```
or
```javascript
const newRoot = join(file, '../../go.mod');
```
Note: `path.join()` retains the relativeness of `file` whereas `path.resolve()` will return an absolute path and in this case, we want a relative path.

Reply:
Interesting! Do you recall the specific issues with splitting on `path.sep`?

I'm for this change, but I'd rather do it separately. I'll keep a note of it. Thanks!

Reply:
I believe this was more of an issue with super old versions of Node.js, but today it is possible to be on a posix machine and doing a `join('foo\\bar', 'baz')` will result in `foo\\bar/baz` and `join(process.cwd(), '\\foo\\bar')` will result in `/path/to/cwd\\foo\\bar`.

Under the hood on Windows, Node converts all `\` to `/`, so on Windows things are surprisingly robust. On posix platforms, `\` is a valid file name character, so Node leaves it alone.

In this case, I don't know if the relative paths returned by `glob()` use platform specific path separators.

I guess the moral of the story is using `path.sep` is no more dangerous than `join()` (on Mac/Linux) and probably more importantly don't mix posix and Windows paths, which the codes doesn't appear to be doing anyways.